### ISSUE TITLE
New buzzwords Azure Durable Functions: Making Serverless Last

### ISSUE THREAD

**Issue by postmillenial** [postmillenial/reading-list#1219]
Azure Durable Functions: Making Serverless Last<br>
http://ift.tt/2zZyyss<br><img width="640" height="426" src="http://ift.tt/2hyofVH" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2017/11/78290fc4-nature.jpg"><p>Serverless can be a confusing name for event-driven computing because it focuses so much on what you don&rsquo;t have to care about rather than what you can do. The name does serve as a reminder that <a href="http://ift.tt/2mctD3P" class="local-link">serverless platforms</a> don&rsquo;t help manage state; and while it&rsquo;s good practice to externalize state so that it doesn&rsquo;t interact with the business logic of an application, all but the simplest applications have state that needs managing.</p>
<p>The new <a href="http://ift.tt/2tNmXuW" class="ext-link" rel="external nofollow">Durable Functions</a> feature in <a href="http://bit.ly/2h30eoP" class="ext-link" rel="external nofollow">Microsoft&rsquo;s</a> own serverless service, <a href="http://ift.tt/1PHShwE" class="ext-link" rel="external nofollow">Azure Functions</a>, is one option for managing state, and <a href="http://ift.tt/21E8ew9" class="ext-link" rel="external nofollow">Chris Anderson</a>, senior program manager for Azure Functions, told the New Stack he hopes to see other frameworks and tools for helping with state so that can stay separate from the business logic. After all, being able to focus on that logic is the real advantage of serverless.</p>
<p>As well as taking away the burden of managing infrastructure, serverless also takes away existing tools for solving problems like state. &ldquo;When we gave you just functions, we took away the lot of things you used to have like frameworks to manage state. This is like trying to create a web server without a framework like ASP.NET so you&rsquo;re having to manage every part of the workflow of each request, only across many different services. Durable Functions is a first step towards doing more than just a point-to-point trigger.&rdquo;</p>
<p>Durable Functions addresses the task of managing state for an application. &ldquo;What customers had to do before was manage all that external state themselves, by messages along on a queue and having a record in a database somewhere with the details of the job, which can be large amounts of data. The queue itself can&rsquo;t hold a whole lot of data; it can only hold a pointer to a place where this is more data.&rdquo;</p>
<h2>Orchestration Patterns</h2>
<p>Durable Functions are intended to address a variety of patterns and scenarios that would quickly get complicated using triggers, timers and notifications, especially when orchestrating a range of activities with a complex set of tasks that need to happen each time a particular event occurs.</p>
<p>The simplest case is a &ldquo;straight line&rdquo; of events. &ldquo;When I have one task, that causes another task to occur, that causes another task to occur, that causes <em>another</em> task to occur&hellip; with some if statements and other business logic to fork the workflow, but I&rsquo;m trying to go from point a to point b,&rdquo; Anderson explained.</p>
<p>Fan-out and fan-in is another common serverless pattern. Think of how MapReduce works. &ldquo;Fanning out work to many parallel nodes works great in serverless because you&rsquo;ve got lots of tiny machines; so if you&rsquo;ve got a big work item, you want to split it up into as many small nodes as possible and fan out, and serverless will keep on fanning out as much in parallel as you need to and then scale back in. Whether you have one big job or lots of big jobs, it right sizes itself so you can always reach maximum values of parallelism,&rdquo; Anderson said.</p>
<p>The part that&rsquo;s always hard is the &ldquo;fan in&rdquo; step, Anderson said, because the workflow needs to know when all those messages have completed, so it can go run the next step of the process. Durable Functions allows the developer to express that with a simple a FOR loop and an AWAIT call to wait for all those things fanned out to complete.</p>
<p>&ldquo;What used to be lots of complex pieces like polling and error cases and so on can all now be expressed in five lines of C#,&rdquo; Anderson boasted. And don&rsquo;t think of that just for data parallel operations; it&rsquo;s also useful for automation, like provisioning multiple VMs in parallel, with multiple steps that have to complete, including having someone sign off on the process.</p>
<p>Durable Functions can keep track of state in looping processes, like a game show with multiple rounds. &ldquo;I can manipulate it to do things that create what I call durable stateful singletons, which is orchestration that runs in a loop that keeps track of the current state of that object, and I can send it external messages to make it do other things.&rdquo;</p>
<p>Another complex scenario Durable Functions can help with is waiting for some external interaction. &ldquo;In the real world, we&rsquo;re going to have other external systems that interact with our orchestrations; I can always try to wrap those things up in other functions but I don&rsquo;t want to have a function that&rsquo;s polling some external service to know whether it&rsquo;s there or not,&rdquo; Anderson said.</p>
<p>Durable Functions supports the idea of an external event, whether that&rsquo;s another service or another function elsewhere, calling into the durable functions to pass it along with a piece of information and Durable Functions can wait on that event.</p>
<h2>State Needn&rsquo;t Be Expensive</h2>
<p>Anderson suggested that would be a good way to handle user interactions with something driven by functions, like a two-factor authentication (2FA) service. &ldquo;Humans don&rsquo;t operate at the same speed as restful services! It&rsquo;s going to take a few seconds to take out the phone to see what the 2FA code is.&rdquo;</p>
<p>The Durable Functions orchestration can send the notification to the user that there&rsquo;s a 2FA code to input, wait for the code to come back from the user, validate it or allow two or three more attempts before killing the login attempt, all without keeping the function running &mdash; and charging you for it.</p>
<p>&ldquo;You don&rsquo;t have long running functions waiting for other things to complete; as soon as it has to make an external call to get state, Durable Functions shuts down the orchestrator and you don&rsquo;t pay for it anymore&nbsp;&mdash; so you only pay for the orchestrator when it&rsquo;s not waiting around for external state.&rdquo;</p>
<p>Durable Functions is obviously a good match for Azure services like <a href="http://ift.tt/2wamYgg" class="ext-link" rel="external nofollow">Event Grid</a> and <a href="http://ift.tt/1zFfVW9" class="ext-link" rel="external nofollow">Event Hubs</a>, but it works well with third-party events too. Anderson mentioned using <a href="http://ift.tt/2tpC0wT" class="local-link">Twilio</a> and a <a href="http://ift.tt/2zYuzwv" class="local-link">webhook</a> request to Functions to add 2FA to Slack without needing to pull out a phone.</p>
<p>&ldquo;Under the covers, it&rsquo;s an HTTP request. You can have another function that acts as the front end for that external event, or the external event can call the admin endpoint directly,&rdquo; Anderson said. &ldquo;In practice, you generally have a function sitting in front and you can then have a couple of different triggers, like a queue or an admin panel with an HTTP endpoint and they can both trigger the function depending on which workflow makes sense.&rdquo;</p>
<p><a href="http://ift.tt/2ah8amS" class="ext-link" rel="external nofollow">Azure Logic Apps</a> workflows have been the main option for orchestration so far (and Durable Functions can run alongside Logic Apps), but that hasn&rsquo;t suited developers who prefer to write code than use a graphical interface.</p>
<p>Currently, workflows can only be composed in C#, but it will cover other languages. Plans are afoot to extend the service to accommodate code written in Node.js, JavaScript, and Typescript.</p>
<p>Although Azure Functions has a persistent file system behind the functions, rather than the usual read-only copy (because it was built upon <a href="http://ift.tt/1JlXXzf" class="ext-link" rel="external nofollow">Azure App Service</a> and <a href="http://ift.tt/2zYuBo7" class="ext-link" rel="external nofollow">WebJobs</a>), but that&rsquo;s not a good place to store state because it&rsquo;s so expensive. Instead, Durable Functions uses Azure storage queues and Azure table storage for key-value pairs. &ldquo;Eventually we&rsquo;ll give you options of using other underlying Azure technologies for higher performance,&rdquo; Anderson said.</p>
<p>&ldquo;We hope to have additional providers that let you move onto using <a href="http://ift.tt/1kfs9kA" class="ext-link" rel="external nofollow">Azure Service Bus</a> queues and <a href="http://ift.tt/2rM7ahR" class="local-link">Cosmos DB</a> for the higher throughput scenarios. Usually, you have to use something like <a href="http://ift.tt/2xzZu4I" class="local-link">Kafka</a> or Event Hubs to manage that kind of throughput,&rdquo; Anderson said.</p>
<p>Theoretically, that abstraction could make it possible for Durable Functions to work locally (as the Functions runtime already can). Whether that happens will depend on customer feedback, but Anderson talked hypothetically about being able to use Redis, for example, as the storage for Durable Functions.</p>
<p>Currently in preview, Durable Functions is likely to move to general availability with version 2 of Azure Functions (which it depends on). As well as the planned expansion of storage options and broader language support, it will add more scaling options in time. &ldquo;We plan on doing auto-sharding for you, so you can take advantage of many instances and get parallel scale through that.&rdquo;</p>
<p>But rather than expanding the scope of Durable Functions, Anderson hopes to see more frameworks and projects aimed at making it easier to do large-scale development with serverless.</p>
<p>&ldquo;We had to make this tradeoff [with serverless]: you don&rsquo;t have to manage infrastructure, but you do have to give up a lot more control over your application pipeline, and there are not many frameworks &mdash; if any &mdash; that work when limited down to a single function entry point. My expectation is that we&rsquo;ll see a new generation of frameworks that operate at the single functions level,&rdquo; Anderson said.</p>
<p class="attribution"><a href="http://bit.ly/2h30eoP" class="ext-link" rel="external nofollow">Microsoft</a> is a sponsor of The New Stack.</p>
<p class="attribution">Feature image <a href="http://ift.tt/2zZXrVb" class="ext-link" rel="external nofollow">via</a> PXhere.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2zZyyss">Azure Durable Functions: Making Serverless Last</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br><br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
November 02, 2017 at 05:00AM
